File: src/main/java/me/andre111/d20server/service/GameService.java
package me.andre111.d20server.service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import me.andre111.d20common.message.game.EnterGame;
import me.andre111.d20common.message.game.EnterMap;
import me.andre111.d20common.message.game.PlayerList;
import me.andre111.d20common.message.game.entity.AddEntity;
import me.andre111.d20common.model.entity.map.Map;
import me.andre111.d20common.model.entity.map.Token;
import me.andre111.d20common.model.entity.profile.Profile;
import me.andre111.d20server.model.EntityManagers;

public abstract class GameService {
	private static java.util.Map<Long, ProfileStatus> joinedProfiles = new HashMap<>();
	
	public static void init() {
		// add atleast one map
		if(EntityManagers.get(Map.class).all().isEmpty()) {
			Map map = new Map("New Map");
			EntityManagers.get(Map.class).add(map);
		}
		
		// reset player states
		joinedProfiles.clear();
	}
	
	public static void joinGame(Profile profile) {
		// leave current game
		if(joinedProfiles.containsKey(profile.id())) leaveGame(profile);

		// update state before joining into game (-> do not recieve messages before loading is finished?)
		updateClientState(profile);
		
		// join new game
		joinedProfiles.put(profile.id(), new ProfileStatus());
	}
	
	public static void updateClientState(Profile profile) {
		// sync data -> moves client into loading state
		EntityManagers.fullSync(profile);
		
		// send enter message -> moves client into main state and sets client role (PLAYER/GM)
		MessageService.send(new EnterGame(profile), profile);
		
		// update players (to all)
		MessageService.send(new PlayerList(UserService.getAllProfiles()), (Map) null);
		
		//TODO: send gamestate (to player) (->LoadMap message, last X chat entries, ...)
		reloadMaps(profile);
		ChatService.sendHistory(profile, 200);
		ChatService.appendNote(profile.getName()+" joined!");
	}
	
	public static void leaveGame(Profile profile) {
		joinedProfiles.remove(profile.id());
		
		// update players
		MessageService.send(new PlayerList(UserService.getAllProfiles()), (Map) null);
		ChatService.appendNote(profile.getName()+" left!");
	}
	
	public static boolean isJoined(Profile profile) {
		return joinedProfiles.containsKey(profile.id());
	}
	
	
	
	public static void reloadMaps(Profile profile) {
		List<Profile> profiles = profile != null ? List.of(profile) : UserService.getAllConnectedProfiles();
		
		// update all players
		for(Profile otherProfile : profiles) {
			Map map = getPlayerMap(otherProfile);
			if(map != null) {
				MessageService.send(new AddEntity(map), otherProfile); // send map because client could have no independent access
				MessageService.send(new EnterMap(map), otherProfile);
			}
		}
	}

	
	public static Map getPlayerMap(Profile profile) {
		long mapID = profile.getCurrentMap();
		return EntityManagers.get(Map.class).find(mapID);
	}
	
	public static Token getSelectedToken(Map map, Profile profile, boolean forceSingle) {
		List<Long> selectedTokens = joinedProfiles.get(profile.id()).selectedTokens;
		if(selectedTokens == null || selectedTokens.isEmpty()) return null;
		if(forceSingle && selectedTokens.size() != 1) return null;
		
		Token token = EntityManagers.get(Token.class).find(selectedTokens.get(0));
		return token;
	}
	public static void setSelectedTokens(Profile profile, List<Long> selectedTokens) {
		joinedProfiles.get(profile.id()).selectedTokens = selectedTokens;
	}
	
	private static final class ProfileStatus {
		private List<Long> selectedTokens = new ArrayList<>();
	}
}

File: core/server/service/game-service.js
// @ts-check
import { fullSync } from '../entity/server-entity-managers.js';
import { EDIT_KEY } from '../handler/filemanager.js';
import { ChatService } from './chat-service.js';
import { MessageService } from './message-service.js';
import { UserService } from './user-service.js';

import { Entity } from '../../common/common.js';
import { Access, Role } from '../../common/constants.js';
import { EntityManagers } from '../../common/entity/entity-managers.js';
import { ChangeConfig, EnterGame, EnterMap, PlayerList } from '../../common/messages.js';
import { ModuleService } from './module-service.js';
import { fromJson, toJson } from '../../common/util/datautil.js';
import { CONFIG } from '../../common/config.js';
import { I18N } from '../../common/util/i18n.js';

export class GameService {
    static init() {
        // add atleast one map
        if (EntityManagers.get('map').all().length == 0) {
            const map = new Entity('map');
            map.setString('name', 'New Map');
            EntityManagers.get('map').add(map);
        }
    }

    static joinGame(profile) {
        GameService.updateClientState(profile);
        ChatService.appendNote(null, I18N.get('game.join', '%0 joined!', profile.getUsername()));
    }

    static leaveGame(profile) {
        MessageService.broadcast(new PlayerList(UserService.getAllProfiles()));
        ChatService.appendNote(null, I18N.get('game.leave', '%0 left!', profile.getUsername()));
    }

    static updateClientState(profile) {
        // sync config
        CONFIG.iterate((key, def, value) => {
            if (def.clientAccessible) {
                MessageService.send(new ChangeConfig(key, value), profile);
            }
        });

        // sync data -> moves client into loading state
        fullSync(profile);

        // send enter message -> moves client into main state ans sets client role/profile and edit key for gms
        MessageService.send(new EnterGame(profile, profile.getRole() == Role.GM ? EDIT_KEY : -1), profile);

        // update player list (to all)
        MessageService.broadcast(new PlayerList(UserService.getAllProfiles()));

        // send game state
        GameService.reloadMaps(profile);
        ChatService.sendHistory(profile, 100);
        ModuleService.sendModuleDefinitions(profile);
    }

    static reloadMaps(profile) {
        // profile can be null => reload for all
        if (profile) {
            const map = EntityManagers.get('map').find(profile.getCurrentMap());
            if (map) {
                EntityManagers.get('map').syncEntity(profile, map); // send map and contained entities because client could have no previous access
                MessageService.send(new EnterMap(map, GameService.getFOW(map, profile)), profile);
            }
        } else {
            UserService.forEach(profile => GameService.reloadMaps(profile));
        }
    }

    static getFOW(map, profile) {
        if (!map || !profile) return [];
        const manager = map.getContainedEntityManager('fow');
        if (!manager) return [];
        const fowEntity = manager.find(profile.getID());
        if (!fowEntity) return [];

        return fromJson(fowEntity.getString('area'));
    }

    static setFOW(map, profile, fow) {
        if (!map || !profile || !fow) throw new Error('Missing required parameter');
        const manager = map.getContainedEntityManager('fow');
        if (!manager.has(profile.getID())) {
            manager.add(new Entity('fow', profile.getID()));
        }

        manager.updateProperties(profile.getID(), { 'area': toJson(fow) }, Access.SYSTEM);
    }

    static resetFOW(map) {
        if (!map) throw new Error('Missing required parameter');
        const manager = map.getContainedEntityManager('fow');
        for (const fowEntity of manager.all()) {
            manager.remove(fowEntity.getID());
        }

        GameService.reloadMaps();
    }
}
